### PR TITLE
feat: allow custom storage using django settings

### PR DESCRIPTION
**Pull request by BryanttV**
### Description
This PR allows you to use custom storage (such as S3) through Django Settings.

### How To Test
1, Install this branch in your environment.
2. Add these settings to your LMS and CMS _(This configuration is for using s3)_
```python
MINDMAP_BLOCK_STORAGE = {
  "storage_class": "storages.backends.s3boto3.S3Boto3Storage",
  "settings": {
      "bucket_name": "your-bucket-name",
  },
}
```
3. Access to the block from Studio and LMS, you should be able to get and upload mind maps normally.

### CODE REVIEW

**Comment by Ian2012** [comment]
I dont think those variables names should be used, you should prefix them to not overwrite the default storage backend settings

**Comment by mariajgrimaldi** [comment]
> I dont think those variables names should be used, you should prefix them to not overwrite the default storage backend settings
Prefix them or use them as if they already exist.

**Comment by mariajgrimaldi** [comment]
We should test this using credentials for S3 and minio

File: mindmap/settings/common.py [hunk 1]
```diff
@@ -10,4 +10,10 @@ def plugin_settings(settings):
     # AWS S3 settings
     settings.AWS_ACCESS_KEY_ID = "CHANGE-ME"
     settings.AWS_SECRET_ACCESS_KEY = "CHANGE-ME"
-    settings.FILE_UPLOAD_STORAGE_BUCKET_NAME = "CHANGE-ME"
+    settings.AWS_DEFAULT_ACL = None
```
Comment (mariajgrimaldi) [inline mindmap/settings/common.py:13]: Be aware of: https://github.com/eduNEXT/xblock-mindmap/pull/8#issuecomment-1666500403

**Review by mariajgrimaldi — CHANGES_REQUESTED** [review]
A few requests :)

**Comment by BryanttV** [comment]
Changes made!

File: mindmap/mindmap.py [hunk 2]
```diff
@@ -254,18 +227,13 @@ def get_current_mind_map(self, path_prefix: str) -> dict | None:
             dict: The current mind map content.
             None: If the file does not exist.
         """
-        s3_client, aws_bucket_name = self.connect_to_s3()
+        mindmap_storage = get_mindmap_storage()
```
Comment (Ian2012) [inline mindmap/mindmap.py:230]: Only load this once before the class definition in a variable `MINDMAP_STORAGE`

File: mindmap/mindmap.py [hunk 3]
```diff
@@ -281,12 +249,11 @@ def upload_file(self, data, _suffix="") -> None:
             user = self.get_current_user()
             data["path_prefix"] = self.anonymous_user_id(user)
```
Comment (Ian2012) [inline mindmap/mindmap.py:250]: Change this `path_prefix` for something more scoped like 'mindmaps_uploads/{{anonymous_user_id}}' or something similar

**Review by Ian2012 — CHANGES_REQUESTED** [review]
Just some improvements, otherwise looks good for me

File: mindmap/mindmap.py (hunk 2 above)
Comment (BryanttV) [reply mindmap/mindmap.py:230]: I tried to load it that way (similar to the H5P xblock), but it didn't work 🤔

File: mindmap/mindmap.py (hunk 3 above)
Comment (BryanttV) [reply mindmap/mindmap.py:250]: Yes, the full path is obtained from the `get_file_key` method.

File: mindmap/mindmap.py (hunk 2 above)
Comment (Ian2012) [reply mindmap/mindmap.py:230]: Then define a class level property with that value

Comment (BryanttV) [reply mindmap/mindmap.py:230]: That doesn't work for me either.

**Comment by mariajgrimaldi** [comment]
While checking the minio interface I found that we are currently using two paths:
```
openedxuploads / mindmaps / 1e6fe8a780a94165a2921e6ff9f501ab / f659f144d1c23d14d98705c8db1666ac / mindmap.json
openedxuploads / mindmaps / f4014e9515dc4d5e9a5c320d3894e1e5 / instructor / mindmap.json
```
If we're going to use `instructor` and `students` to categorize the mindmaps, I believe this path would be more organized:
```
openedxuploads / mindmaps / students / 1e6fe8a780a94165a2921e6ff9f501ab / f659f144d1c23d14d98705c8db1666ac / mindmap.json
openedxuploads / mindmaps / instructor / f4014e9515dc4d5e9a5c320d3894e1e5  / mindmap.json
```
So I can access students/instructors mindmaps without figuring out first the IDs.

**Comment by BryanttV** [comment]
To implement [this change](https://github.com/eduNEXT/xblock-mindmap/pull/8#issuecomment-1671623807) the `get_file_key` method could be modified like this, what do you think? @mariajgrimaldi 

```python
def get_file_key(self, path_prefix: str) -> str:
        """
        Return the key (path) to save and retrieve the file in S3.

        Args:
            path_prefix (str): The path prefix to use in the key (path).

        Returns:
            str: The key (path) to use in S3.
        """
        block_id = self.scope_ids.usage_id.block_id

        if path_prefix == "instructor":
            return f"mindmaps/instructors/{block_id}/mindmap.json"

        return f"mindmaps/students/{block_id}/{path_prefix}/mindmap.json"
```

**Comment by mariajgrimaldi** [comment]
@BryanttV: I think you meant `anonymous_user_id` instead of the second `path_prefix`?

File: mindmap/mindmap.py (hunk 2 above)
Comment (Ian2012) [reply mindmap/mindmap.py:230]: Then set it as a Django setting or leave it like it if it doesn't work

**Comment by BryanttV** [comment]
The `path_prefix` is used throughout the flow, and its value can be `"instructor"` or the anonymous user id. Perhaps it might be a more appropriate name to leave it as `anonymous_user_id`, which can take the value of `None` _(if it is an instructor)_, or the anonymous user id _(if it is a student)_. @mariajgrimaldi

File: mindmap/mindmap.py (hunk 2 above)
Comment (BryanttV) [reply mindmap/mindmap.py:230]: Adding a new Django setting I think is not necessary, I will leave it as it is now.

**Review by Ian2012 — APPROVED** [review]

**Review by mariajgrimaldi — APPROVED** [review]
We're ready to go since the paths are working as expected! Remember to update the changelog entries and to bump the xblock version :)

**Pull request merged**